text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,261 @@
+and a new `thread::scoped` API that takes advantage of it.
+
+## Deferred computation
+
+Here's a very simple API for deferring computation:
+
+```
+// in std:
+
+mod thread {
+    pub struct Scope<'a> { ... }
+
+    pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R;
+
+    impl<'a> Scope<'a> {
+        pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a;
+    }
+}
+```
+
+(This is put in the `std::thread` module because (1) scopes are per-thread
+concepts (2) it fits nicely with other functionality like `panicking` and
+`catch_panic`.)
+
+You call `scope` to introduce a new `Scope` value, which is passed into a
+closure that is immediately invoked.
+
+The closure can use the `defer` method to register callbacks to invoke upon
+*any* exit from the callback, including unwinding. Unlike RAII guards, there is
+no way to leak these callbacks, and the implementation shown below works around
+[known cases](https://github.com/rust-lang/rust/issues/14875) of destructor
+leakage.
+
+This is a generally useful mechanism that can avoid the need to create custom
+RAII guards for situations where you might use `try`/`finally` in other
+languages. But it is also just the support needed for scoped threads.
+
+## Scoped threads
+
+To recover scoped threads, we extend the `Scope` type with a method for spawning
+threads:
+
+```rust
+impl<'a> Scope<'a> {
+    pub fn spawn<F, T>(&self, f: F) -> thread::JoinHandle<T> where
+        F: FnOnce() -> T + Send + 'a,
+        T: Send + 'a;
+}
+```
+
+Like the original `thread::scoped` API, this allows for the child thread's
+closure to borrow data (with lifetime `'body`) from the parent thread. These
+borrows are bounded by the lifetime of the `Scope` value, and the implementation
+uses the `defer` method to add a callback that will join on (wait for completion
+of) the child thread on exit from the scope -- thus restoring memory safety.
+
+Note that, while previously one might return `JoinGuard`s outward to expand the
+scope of joining, the pattern here is reversed: you call `scope` at the
+outer-most scope, and then pass a reference to the `Scoped` value inward.
+
+Putting it all together, here's an example from TRPL, and a version adjust to the new API:
+
+```rust
+// using the thread::scoped API
+fn old_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    let guards: Vec<_> = (0..2).iter().map(|_| {
+        thread::scoped(|| {
+            let mut data = data.lock().unwrap();
+            data[i] += 1;
+        })
+    }).collect();
+
+    // threads implicitly joined here, when `guards` goes out of scope and drops
+    // its contents
+}
+```
+
+```rust
+// using the proposed thread::scope API
+fn new_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    thread::scope(|s| {
+        for i in 0..2 {
+            s.spawn(|| {
+                let mut data = data.lock().unwrap();
+                data[i] += 1;
+            });
```

Comment:
@aturon: That example only works (after fixing the attribute to be a crate attribute and adding `use std::sync::Mutex;`) because you're dropping the guard inside of the `for` loop. It is in fact borrowing the `i` value, but the borrow expires inside of the loop. So your code is actually not being parallel at all (which seems like an easy mistake to make with that RAII pattern). Rewriting it to do the loop in parallel:

```rust
#![feature(scoped)]
use std::thread;
use std::sync::Mutex;

fn main() {
    let data = Mutex::new(vec![1u32, 2, 3]);

    let mut guards = vec![];
    for i in 0..3 {
        guards.push(thread::scoped(|| {
            let mut data = data.lock().unwrap();
            data[i] += 1;
        }));
    }
    drop(guards);

    println!("{:?}", data);
}
```

yields the following error:

> ```
unnamed.rs:10:36: 13:10 error: `i` does not live long enough
unnamed.rs:10         guards.push(thread::scoped(|| {
unnamed.rs:11             let mut data = data.lock().unwrap();
unnamed.rs:12             data[i] += 1;
unnamed.rs:13         }));
unnamed.rs:6:45: 18:2 note: reference must be valid for the block suffix following statement 0 at 6:44...
unnamed.rs:6     let data = Mutex::new(vec![1u32, 2, 3]);
unnamed.rs:7
unnamed.rs:8     let mut guards = vec![];
unnamed.rs:9     for i in 0..3 {
unnamed.rs:10         guards.push(thread::scoped(|| {
unnamed.rs:11             let mut data = data.lock().unwrap();
              ...
unnamed.rs:9:9: 9:10 note: ...but borrowed value is only valid for the for at 9:8
unnamed.rs:9     for i in 0..3 {
                     ^
```